### DISCUSSION TITLE
D/B bit in data selectors and `ss`

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1609] [General]
I noticed two oddities in dosemu2 lately, while creating tests for lDebug and FreeDOS Debug.

The first is that a selector's D/B bit (which indicates 32-bit code or stack size when set) is by default set by dosemu2. On HDPMI32 It is clear by default. [Here is a test case](https://hg.pushbx.org/ecm/dpmitest/file/08e80e80f427/dpmioffs.asm). On dosemu2 it outputs:

```
Welcome to dosemu2!
    Build 2.0pre8-20191102-1332-g9cf2b7560
About to Execute : dpmioffs.com
About to Execute : dpmioffs.com
Protected mode breakpoint at 015Dh.
32-bit code segment breakpoint at 016Fh.
32-bit data selector (D bit 0) breakpoint at 01F3h.
32-bit data selector (D bit 1) breakpoint at 0216h.

Welcome in 32-bit protected mode.
Check null selector: The D bit is indeterminate.
Check allocated:     The D bit is set (1).
Check set base:      The D bit is set (1).
Check set limit:     The D bit is set (1).
Check D bit cleared: The D bit is clear (0).
Check D bit set:     The D bit is set (1).
```

The second is that dosemu2 appears to sometimes assume that the `ss` B bit is set (running in a 32-bit `cs` in DPMI). [Here is the next test case](https://hg.pushbx.org/ecm/dpmitest/file/08e80e80f427/dpmistck.asm). Output when run on its own, which is as expected:

```
Welcome to dosemu2!
    Build 2.0pre8-20191102-1332-g9cf2b7560
About to Execute : dpmistck.com
About to Execute : dpmistck.com
Protected mode breakpoint at 0211h.
32-bit code segment breakpoint at 0223h.
16-bit stack test breakpoint at 031Bh.
32-bit stack test breakpoint at 0339h.
Return from child process at 0450h.

Welcome in 32-bit protected mode.
Check allocated:     The D bit is set (1).
Check set limit:     The D bit is set (1).
Check set base:      The D bit is set (1).
DPMI allocation at 02121000h.
DPMI allocation selector is 00EFh.
Check D bit cleared: The D bit is clear (0).
ESP = 0001FFFCh, Trace Flag appears as clear (0).
Check D bit set:     The D bit is set (1).
ESP = 0000FFFCh, Trace Flag appears as clear (0).
Child process terminated okay, back in real mode.
```

Note that the first `ESP = ` line indicates that the 16-bit `sp` underflowed without affecting the upper word of `esp`. This is the expected behaviour.

However, when tracing the 16-bit stack test with lDebugX then the full `esp` is used even though the B bit is clear:

```
Welcome to dosemu2!
    Build 2.0pre8-20191102-1332-g9cf2b7560
About to Execute : ldebugx.com /c='r dco or= 800' dpmistck.com
About to Execute : ldebugx.com /c='r dco or= 800' dpmistck.com
-r dco or= 800
-g
DPMI entry hooked, new entry=046C:8A66
Protected mode breakpoint at 0211h.
32-bit code segment breakpoint at 0223h.
16-bit stack test breakpoint at 031Bh.
32-bit stack test breakpoint at 0339h.
Return from child process at 0450h.

Unexpected breakpoint interrupt
AX=0200 BX=00E6 CX=0005 DX=000A SP=FFFC BP=0000 SI=070E DI=8A66
DS=220F ES=3210 SS=220F CS=220F IP=01D3 NV UP EI PL ZR NA PE NC
220F:01D3 89E5              mov     bp, sp
-rx
386 registers are displayed
-g 31B
Succeeded in unhooking interrupt 2Fh.
Welcome in 32-bit protected mode.
Check allocated:     The D bit is set (1).
Check set limit:     The D bit is set (1).
Check set base:      The D bit is set (1).
DPMI allocation at 02121000h.
DPMI allocation selector is 012Fh.
Check D bit cleared: The D bit is clear (0).
EAX=00000200 EBX=0000012F ECX=0001F300 EDX=0212100A ESP=0000FFFC EBP=0000FFFC
ESI=0000074C EDI=00000006 NV UP EI PL ZR NA PE NC
DS=00AF ES=00AF SS=012F CS=0107 FS=0000 GS=0000 EIP=0000031B
0107:0000031B 8ED3          mov     ss, ebx
0107:0000031D BC00000100    mov     esp, 00010000
#t
EAX=00000200 EBX=0000012F ECX=0001F300 EDX=0212100A ESP=00010000 EBP=0000FFFC
ESI=0000074C EDI=00000006 NV UP EI PL ZR NA PE NC
DS=00AF ES=00AF SS=012F CS=0107 FS=0000 GS=0000 EIP=00000322
0107:00000322 50            push    eax
#
EAX=00000200 EBX=0000012F ECX=0001F300 EDX=0212100A ESP=0000FFFC EBP=0000FFFC
ESI=0000074C EDI=00000006 NV UP EI PL ZR NA PE NC
DS=00AF ES=00AF SS=012F CS=0107 FS=0000 GS=0000 EIP=00000323
0107:00000323 89E1          mov     ecx, esp
#
EAX=00000200 EBX=0000012F ECX=0000FFFC EDX=0212100A ESP=0000FFFC EBP=0000FFFC
ESI=0000074C EDI=00000006 NV UP EI PL ZR NA PE NC
DS=00AF ES=00AF SS=012F CS=0107 FS=0000 GS=0000 EIP=00000325
0107:00000325 58            pop     eax
#
EAX=00000200 EBX=0000012F ECX=0000FFFC EDX=0212100A ESP=00000000 EBP=0000FFFC
ESI=0000074C EDI=00000006 NV UP EI PL ZR NA PE NC
DS=00AF ES=00AF SS=012F CS=0107 FS=0000 GS=0000 EIP=00000326
0107:00000326 9C            pushf
#
EAX=00000200 EBX=0000012F ECX=0000FFFC EDX=0212100A ESP=0000FFFC EBP=0000FFFC
ESI=0000074C EDI=00000006 NV UP EI PL ZR NA PE NC
DS=00AF ES=00AF SS=012F CS=0107 FS=0000 GS=0000 EIP=00000327
0107:00000327 5F            pop     edi
#
EAX=00000200 EBX=0000012F ECX=0000FFFC EDX=0212100A ESP=00000000 EBP=0000FFFC
ESI=0000074C EDI=00000246 NV UP EI PL ZR NA PE NC
DS=00AF ES=00AF SS=012F CS=0107 FS=0000 GS=0000 EIP=00000328
0107:00000328 0FB225780400
              00            lss     esp, [00000478]   DS:00000478=012F:0000FFFC
#
```

The `esp` value as saved in `ecx` indicates that the full 32-bit `esp` waa decremented.

**Comment by ecm-pushbx** [comment]
Hmm, odd. I just noticed that the 16-bit stack test in the debugger does increment `esp` from `0000FFFC` to `00000000` upon the `pop eax` instruction. If it used the full `esp` consistently we would expect `00010000`.

**Reply by stsp** [reply]
Please attach binary test-cases.
Is this with cpu-emu or kvm?

**Reply by stsp** [reply]
I suspect high ESP word was just
zeroed, rather than been the subject
of decrementing.
Please don't misinterpret your results. :)

**Reply by ecm-pushbx** [reply]
> I suspect high ESP word was just zeroed,

Oh true that might be so. Still odd that it happens when tracing in the debugger but not otherwise.

Test binaries are at https://pushbx.org/ecm/test/20220329/

How do I tell whether it is cpu-emu or kvm? The content of `/proc/cpuinfo` does list `svm` in the `flags` list.

**Comment by stsp** [comment]
> Oh true that might be so. Still odd that it happens when tracing in the debugger but not otherwise.

Try `$_cpu_vm_dpmi = "native"` and see
if something changes.

> How do I tell whether it is cpu-emu or kvm?

By looking into ~/.dosemu/boot.log.

**Comment by stsp** [comment]
As for D bit: it simply matches
the bitness of a client, and DPMI
spec doesn't specify that explicitly.

**Comment by ecm-pushbx** [comment]
> Try `$_cpu_vm_dpmi = "native"` and see if something changes.

No change with `-I "cpu_vm_dpmi native"`. With either `-I "cpu_vm_dpmi kvm"` (NB I found that there is no file `/dev/kvm` on this machine!) or `-I "cpu_vm_dpmi emulated"` the behaviour changes, both in the same way:

1. The high word of `esp` is cleared even when not singlestepping.
2. The `dosemu -dumb` screen output is interspersed by a bunch of messages reading `ERROR: touched jit-protected page` without a CR (LF only).
3. Upon termination of the dpmistck processes, lDebugX thinks that it is in 86 Mode but with the InDOS flag set. This is usually due to retaining the Protected Mode pointer to the flag, ie the debugger did not properly process the switch back to 86 Mode. I would have to inspect this further to be sure however.

> As for D bit: it simply matches the bitness of a client, and DPMI spec doesn't specify that explicitly.

Allright, I suspected it may be underspecified. The [new test "HUGESEG"](https://github.com/Baron-von-Riedesel/DOS-debug/blob/b60a75d445f08f9a20abc48ccfb906225194bb41/Test/HUGESEG.ASM#L49-L64) of the original DebugX depends on the HDPMI32 behaviour for its intended purpose. However, I don't think dosemu2 has to do this the same way.

**Reply by stsp** [reply]
Please test the patch here:
https://github.com/dosemu2/dosemu2/pull/1613
about ESP.
I can't reproduce "touched jit-protected page",
please open a bug for that.

**Reply by ecm-pushbx** [reply]
Running this patch with `cpu_vm_dpmi emulated` works as expected. With `cpu_vm_dpmi native` it still zeroes the high word, and when trying to terminate a DPMI process dosemu starts to spew an endless stream of `ERROR: MAP_FIXED_NOREPLACE doesn't work`.

I cannot reproduce the other errors on a newer dosemu2 revision either. Neither the "touched jit-protected page" nor the incorrect segment/selector mixup that leads lDebugX to think it is InDOS.

(I had to rebuild dosemu because its dosdebug used a library that got removed on the last server upgrade. However, even with the rebuilt dosdebug/dosemu2 pair (of the old revision) the InDOS error disappears when tracing through the installdpmi function.)

**Reply by ecm-pushbx** [reply]
This command is sufficient to trigger the endless repetition of the `ERROR: MAP_FIXED_NOREPLACE doesn't work` error message: `dpmitest$ ~/local.new/bin/dosemu -K "$PWD" -dumb -E "dpmistck.com" -I "cpu_vm_dpmi native"`

**Reply by stsp** [reply]
No, can't trigger that and not
confirming that "native" has that
problem.
What kernel is that, where nothing
works?

**Reply by stsp** [reply]
xen?

**Reply by stsp** [reply]
We do not support xen. :)

**Reply by ecm-pushbx** [reply]
Not sure,  `uname -a` gives `Linux ulukai.org 4.19.0 #1 SMP Thu Sep 23 18:26:47 MSK 2021 x86_64 GNU/Linux` though I am unsure whether the kernel version is accurate.

**Reply by stsp** [reply]
What hypervisor is running?
xen?

**Reply by ecm-pushbx** [reply]
Partner says openvz likely.

**Reply by ecm-pushbx** [reply]
Checked out the most recent revision before today's patch and it has the same endless error problem.

**Reply by stsp** [reply]
Please do this:
```
diff --git a/src/arch/linux/mapping/mapping.c b/src/arch/linux/mapping/mapping.c
index aa3339d21..0a52db04a 100644
--- a/src/arch/linux/mapping/mapping.c
+++ b/src/arch/linux/mapping/mapping.c
@@ -417,7 +417,7 @@ static void *do_mmap_mapping(int cap, void *target, size_t mapsize, int protect)
 #ifdef __linux__
     /* under valgrind this flag doesn't work */
     if (kernel_version_code >= KERNEL_VERSION(4, 17, 0) && addr != target)
-      error("MAP_FIXED_NOREPLACE doesn't work\n");
+      dosemu_error("MAP_FIXED_NOREPLACE doesn't work\n");
 #endif
 #endif
     if (addr != target) {
```

And attach the log here.

**Reply by ecm-pushbx** [reply]
Fills up the log repeatedly with memory layout and debugging info. https://pushbx.org/ecm/test/20220329/boot.log I had to kill the dosemu.bin process.

**Reply by stsp** [reply]
> when trying to terminate a DPMI process dosemu starts to spew an endless stream of 

How can that be if the log clearly
shows that this happens on dosemu's
start-up?

**Reply by stsp** [reply]
In either case I disabled the
offending code. No guarantee
it will work on your setup in native
mode now, but worth a try.

**Reply by ecm-pushbx** [reply]
> How can that be if the log clearly shows that this happens on dosemu's start-up?

It didn't happen that way earlier.

> In either case I disabled the offending code. No guarantee it will work on your setup in native mode now, but worth a try.

Doesn't spew the message anymore. However, now I get the following error exit from dosemu upon calling the DPMI entrypoint:

```
Welcome to dosemu2!
    Build 2.0pre9-dev-20220330-113-gfbc50bfbe
Protected mode breakpoint at 0211h.
32-bit code segment breakpoint at 0223h.
16-bit stack test breakpoint at 031Bh.
32-bit stack test breakpoint at 0339h.
Return from child process at 0450h.

Security warning: native DPMI mode is insecure, adjust $_cpu_vm_dpmi
ERROR: KVM: error opening /dev/kvm: No such file or directory
ERROR: PKT: Cannot open tap0: Permission denied
ERROR: PKT: Cannot open tap0: Permission denied
ERROR: alsa_midi (ALSA lib): seq_hw.c:466:(snd_seq_hw_open) : No such file or directory open /dev/snd/seq failed
ERROR: Unable to open console or check with X to evaluate the keyboard map.
Please specify your keyboard map explicitly via the $_layout option. Defaulting to US.
ERROR: Your kernel does not support UC_STRICT_RESTORE_SS and the work-around in dosemu is not enabled.
ERROR: SS selector invalid: 0x002B, type=0 np=0
ERROR: leavedos called recursively, forgetting the graceful exit!
removing stale files:
[home]/.dosemu/run/dosemu.dbgin.1035493
[home]/.dosemu/run/dosemu.dbgout.1035493
```

**Reply by stsp** [reply]
Should now be fixed.

**Reply by ecm-pushbx** [reply]
Upon the test program entering PM I get a lot of output, including the `UC_STRICT_RESTORE_SS` message. Look at https://pushbx.org/ecm/test/20220330.txt

Partner says that the kernel running is actually 3.10.0 (on 
Virtual Server 8.3), according to https://www.hosteurope.de/faq/server/virtual-server/kernel-version/ -- Is there a way to override the kernel version dosemu sees?

**Reply by stsp** [reply]
Enlarged stack buffer.
Please see if the stack overflow is still there.
I don't think any override is needed.
I think 3.10 is running on a machine,
and 4.19 (which dosemu is seeing) is
running under hypervisor, that is running
on top of 3.10.

**Reply by ecm-pushbx** [reply]
Displays less spam but still crashes upon entry into PM. Output is listed in https://pushbx.org/ecm/test/20220330.2

Partner says they use a 3.10 based kernel for the container host *and* guest (insofar as you can call it that, it's just a kernel running containers with different userlands), which has 4.19 level features probably backported but might be incomplete / different I think.

The link I gave from hosteurope says:

> Es kommen speziell für die Virtualisierung angepasste Kernel zum Einsatz, die auch unabhängig vom sonstigen Kernel gepflegt werden. Bitte beachten Sie, dass aufgrund des Virtualisierungsverfahrens nicht alle Funktionen, die der Standardkernel unterstützt, auch innerhalb der virtuellen Maschine genutzt werden können und andererseits auch Funktionen aus neueren Kernels, die z.B. für den Betrieb neuere OS nötig sind, in den älteren Kernelversionen eingepflegt werden. Dadurch kommt es auch dazu, dass in Ihrem System ggf. eine andere/neuere Kernelversion angezeigt wird, als vom Wirt tatsächlich genutzt wird.

My translation:

> There are kernels especially adapted to virtualisation being used, that are also cared for independent of other kernels. Please note that due to the virtualisation method not all functions supported by the standard kernel can be used in the virtual machine, though on the other hand functions from newer kernels, that are eg needed to run newer OS, are being added into the older kernel versions. Due to this, in your system there may be displayed a different/newer kernel version than is actually being used by the host.

**Reply by stsp** [reply]
And now?

**Reply by ecm-pushbx** [reply]
Still crashes, a bit more verbose again: https://pushbx.org/ecm/test/20220330.3

**Reply by stsp** [reply]
Well, it tells you "please install gdb!"
and you do nothing. Not sure what it
should do, probably scream from the
speakers...

Anyway, please try setting WANT_SIGRETURN_WA
to 1 in include/emu.h

**Reply by ecm-pushbx** [reply]
gdb is and was already installed:

```
$ gdb
GNU gdb (Debian 10.1-1.7) 10.1.90.20210103-git
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb)
```

Editing the `src/include/emu.h` to change that define to 1 then re-running `make` outputs the following while compiling `signal.c`: (Not sure whether it was there before, caught it by accident.)

```
signal.c:55:10: warning: #warning Not disabling SIGALTSTACK_WA, update your kernel
[-Wcpp]
   55 |         #warning Not disabling SIGALTSTACK_WA, update your kernel
      |          ^~~~~~~
signal.c:90:10: warning: #warning Not disabling SIGRETURN_WA, update your kernel [-
Wcpp]
   90 |         #warning Not disabling SIGRETURN_WA, update your kernel
      |          ^~~~~~~
```

The stack DPMI test works now:

```
dpmitest$ ~/local.new/bin/dosemu -K "$PWD" -dumb -E "dpmistck.com" -I "cpu_vm_dpmi native"
[...]
Welcome to dosemu2!
    Build 2.0pre9-dev-20220330-117-g84a8625f8
Protected mode breakpoint at 0211h.
32-bit code segment breakpoint at 0223h.
16-bit stack test breakpoint at 031Bh.
32-bit stack test breakpoint at 0339h.
Return from child process at 0450h.

Welcome in 32-bit protected mode.
Check allocated:     The D bit is set (1).
Check set limit:     The D bit is set (1).
Check set base:      The D bit is set (1).
DPMI allocation at 02122000h.
DPMI allocation selector is 00FFh.
Check D bit cleared: The D bit is clear (0).
ESP = 0001FFFCh, Trace Flag appears as clear (0).
Check D bit set:     The D bit is set (1).
ESP = 0000FFFCh, Trace Flag appears as clear (0).
Child process terminated okay, back in real mode.
Security warning: native DPMI mode is insecure, adjust $_cpu_vm_dpmi
ERROR: KVM: error opening /dev/kvm: No such file or directory
ERROR: PKT: Cannot open tap0: Permission denied
ERROR: PKT: Cannot open tap0: Permission denied
ERROR: alsa_midi (ALSA lib): seq_hw.c:466:(snd_seq_hw_open) : No such file or direc
tory open /dev/snd/seq failed
ERROR: MAP_FIXED_NOREPLACE doesn't work
ERROR: Unable to open console or check with X to evaluate the keyboard map.
Please specify your keyboard map explicitly via the $_layout option. Defaulting to
US.
```

NB there is single "ERROR: MAP_FIXED_NOREPLACE doesn't work" now but it doesn't endlessly repeat. And the high word of `esp` is not zeroed.

However, trying to debug the test with lDebugX failed one time when trying to display the 32-bit register dump, output is in https://pushbx.org/ecm/test/20220330.4 However this is not reliably reproduceable. It does sometimes happen, seemingly at random.

Otherwise, when tracing the execution with lDebugX I get `esp` high word as zero again. When proceeding past the test (using `G` command) it does preserve the high word.

Furthermore the debugger (in `dosemu -dumb` connected to DOS stdin/stdout) now repeats input lines once Enter is pressed, eg the prompt is `-` then entering `rx<Enter>` it displays:

```
-rx
-rx
```

And the up and down arrow keys don't arrive into the DOS stdin in the way expected by the debugger. This is all handled by [lineio.asm](https://hg.pushbx.org/ecm/ldebug/file/release4/source/lineio.asm), particularly in `rawinput` and `getc`. Entering `<Up><Enter>` at an empty prompt results in this output now:

```
-^[[A
-
-
```

Also, entering Control-C aborts the dosemu process with the last error output message being `Terminated with signal 2`

**Reply by ecm-pushbx** [reply]
Note that the debugger is started with `/c='r dco or= 800'` which makes it so `rawinput` and `getc` are used. The DOS interrupt 21h service 0Ah call is **not** used then. This enabled using the debugger's line edit and history features in `dosemu -dumb` mode previously.

**Reply by stsp** [reply]
Please try again (w/o any mods).

I don't think you are attaching the log.
I think you are copy/pasting the screen
output instead!

If you need some non-dumb features,
then do not use -dumb. Use `-td -kt`
instead (compatible with any dosemu).

**Reply by ecm-pushbx** [reply]
Made a new issue for these crashes: https://github.com/dosemu2/dosemu2/issues/1615

**Reply by ecm-pushbx** [reply]
I also made a `-dumb` issue: https://github.com/dosemu2/dosemu2/issues/1614

**Comment by ecm-pushbx** [comment]
My partner confirms that this server doesn't have KVM available for our machine.
